controller/disk_controller.go

```diff
@@ -0,0 +1,541 @@
+				dc.eventRecorder, disk, v1.EventTypeWarning)
+		}
+		if disk.Status.State == types.DiskStateConnected {
+			disk.Status.Conditions = types.SetConditionAndRecord(disk.Status.Conditions,
+				types.DiskConditionTypeReady, types.ConditionStatusTrue,
+				"", "",
+				dc.eventRecorder, disk, v1.EventTypeNormal)
+		}
+	}()
+
+	disk.Status.State = types.DiskStateConnected
+	if disk.Spec.NodeID == "" {
+		diskFailureReason = string(types.DiskConditionReasonNodeUnknown)
+		diskFailureMessage = "The Node ID is empty"
+		return nil
+	}
+	isNodeDown, err := dc.ds.IsNodeDownOrDeleted(disk.Spec.NodeID)
+	if err != nil {
+		return err
+	}
+	if isNodeDown {
+		diskFailureReason = string(types.DiskConditionReasonNodeUnknown)
+		diskFailureMessage = fmt.Sprintf("The connected node %v is down or deleted", disk.Spec.NodeID)
+		return nil
+	}
+	// Prevent the corner case:
+	// The corresponding ndoe is up but the ownership of the disk hasn't been transferred.
+	// In this case, Longhorn cannot verify UUID and FSID for the disk.
+	if disk.Spec.NodeID != dc.controllerID {
+		return nil
+	}
+
+	info, err := dc.getDiskInfoHandler(disk.Spec.Path)
+	if err != nil {
+		diskFailureReason = string(types.DiskConditionReasonNoDiskInfo)
+		diskFailureMessage = fmt.Sprintf("Get disk information, error: %v", err)
+		return nil
+	}
+
+	// Check disks in the same filesystem
+	if disk.Status.FSID == "" {
+		isDuplicate, err := dc.isFSIDDuplicatedWithExistingReadyDisk(disk, info.Fsid)
+		if err != nil {
+			return err
+		}
+		// Found multiple disks in the same Fsid
+		if isDuplicate {
+			diskFailureReason = string(types.DiskConditionReasonDiskFilesystemChanged)
+			diskFailureMessage = fmt.Sprintf("Disk has same file system ID %v as other disks", info.Fsid)
+			return nil
+		}
+		disk.Status.FSID = info.Fsid
+	} else {
+		if disk.Status.FSID != info.Fsid {
+			diskFailureReason = string(types.DiskConditionReasonDiskFilesystemChanged)
+			diskFailureMessage = fmt.Sprintf("Recorded file system ID %v is different from the detected file system ID %v", disk.Status.FSID, info.Fsid)
+			return nil
+		}
+	}
+
+	diskUUID := ""
+	diskConfig, err := dc.getDiskConfig(disk.Spec.Path)
+	if err != nil {
+		if !types.ErrorIsNotFound(err) {
+			diskFailureReason = string(types.DiskConditionReasonNoDiskInfo)
+			diskFailureMessage = fmt.Sprintf("Failed to get disk config: %v", err)
+			return nil
+		}
+	} else {
+		diskUUID = diskConfig.DiskUUID
+	}
+	if disk.Status.DiskUUID == "" {
+		if diskUUID == "" {
+			diskConfig, err := dc.generateDiskConfig(disk.Spec.Path)
+			if err != nil {
+				diskFailureReason = string(types.DiskConditionReasonNoDiskInfo)
+				diskFailureMessage = fmt.Sprintf("Failed to generate disk config: %v", err)
+				return nil
+			}
+			diskUUID = diskConfig.DiskUUID
+			log.Debugf("New disk is introduced for Longhorn, a disk config file with UUID %v will be generated", diskUUID)
+		}
+		if disk.Status.DiskUUID != diskUUID {
+			log.Debugf("Set disk UUID to %v", diskUUID)
```

Comment:
Is this an existing disk?

Reply:
This means the disk hasn't been used as a Longhorn disk. And Longhorn needs to generate UUID for it.